### PR TITLE
Improved debugger

### PR DESCRIPTION
**Pull request by FrancescoZini**
The debugger is now split in 2 threads: UI and processor. The ui simply display what the processor does, in this way moving the window of the debugger around will feel smoother as it is handled separately from the processing.

I've also added a new feature to create new filters for the colors to be displayed next to the combined filtered image. This way should be easier to create a new filter (you can test it by setting for example the 'gold' values on the trackbars and then the single-filter image will display only gold.

I've made this change because I found it useful to create some filters.

@aeon0  would you mind giving it a try and letting me know your thoughts? When you got a chance of course, I wrote this quite quickly this morning  so I just raised this as a Draft

The reason I'm asking is because I would like to change a bit the UI of the debugger to be able to load color filters from the config to a text box but that requires using another library for the ui that is not opencv as it does not support plain text, so my question is more with regards of how ok are you with people adding new libraries to the project

### CODE REVIEW

File: src/game_controller.py [hunk 1]
```diff
@@ -108,4 +107,4 @@ def stop(self):
         GameController.is_running = False
 
     def toggle_pause_bot(self):
-        self.bot.toggle_pause()
+        if self.bot: self.bot.toggle_pause()
```
[the remark this answers is not in the document]
Comment (FrancescoZini) [reply src/game_controller.py:110]: did so, I'm reworking the ui to be easier to use to create new layers, add and remove them and see the newly rendered image as it happens (and potentially mix&match different color schemas layers), I'd rather not merge these changes for the time being, I should be done in the next couple of days depending on my RL workload

**Comment by FrancescoZini** [comment]
@aeon0 not draft anymore. I've added a better UI to allow for the creation of filtering layers.

I would suggest to download the code and try it out beacuse just by looking at the code might look a bit convoluted to understand as it is a bit UI-ish.

I can help you understand what's doing it if you want but the general idea is the following:
F10 will start the new debugger (if set in the config - params.ini - defaulted to False)
The debugger will be preloaded with the current combined filters that are in the config (marked as existing layers)
The debugger will be preloaded with the item_filtering (to see whether it matches templates or not)
You can now remove the existing filters and start creating a new one, these are the steps:
 1. give a name to your filter
 2. tick the 'Filter with the current layer only' box
 3. move the slidebars to adjust your filter
 4. Once you are happy with your layer press 'Add current layer to active'
 5. repeat from 1

if you want to see all the layers you created in bitwise_or on the screen just untick the box 'Filter with the current layer only'
if you want to see whether the current item filtering works with your new filter enable the Item Filter by checking its box

**Comment by aeon0** [comment]
Looks good. I think by merging your fix pr there are again some merge conflicts. Ping me once you resolved them and we can merge.

**Comment by FrancescoZini** [comment]
> Looks good. I think by merging your fix pr there are again some merge conflicts. Ping me once you resolved them and we can merge.

I pulled from master, merged and pushed, should be ok now.

**Review by aeon0 — APPROVED** [review]

**Pull request merged**